checker/src/org/checkerframework/checker/index/upperbound/UpperBoundVisitor.java

```diff
@@ -142,6 +305,94 @@ private void visitAccess(ExpressionTree indexTree, ExpressionTree arrTree) {
+    }
+
+    @Override
+    protected void commonAssignmentCheck(
+            Tree varTree, ExpressionTree valueExp, /*@CompilerMessageKey*/ String errorKey) {
+        super.commonAssignmentCheck(varTree, valueExp, errorKey);
+        if (varTree.getKind() == Kind.VARIABLE) {
+            // Not a reassignment, so nothing to check
+            return;
+        }
+        Element elt = InternalUtils.symbol(varTree);
+
+        if (elt != null
+                && (elt.getKind() == ElementKind.FIELD
+                        || elt.getKind() == ElementKind.PARAMETER
+                        || elt.getKind() == ElementKind.LOCAL_VARIABLE)
+                && !ElementUtils.isEffectivelyFinal(elt)) {
+            TypeKind typeKind = InternalUtils.typeOf(varTree).getKind();
+            Receiver field = FlowExpressions.internalReprOf(atypeFactory, (ExpressionTree) varTree);
+            SideEffectKind kind;
+            if (elt.getKind() == ElementKind.FIELD) {
+                if (typeKind == TypeKind.ARRAY) {
+                    kind = UpperBoundUtil.SideEffectKind.ARRAY_FIELD_REASSIGNMENT;
+                } else if (!typeKind.isPrimitive()) {
+                    kind = UpperBoundUtil.SideEffectKind.NON_ARRAY_FIELD_REASSIGNMENT;
+                } else {
+                    return;
+                }
+            } else if (typeKind == TypeKind.ARRAY) {
+                kind = SideEffectKind.LOCAL_VAR_REASSIGNMENT;
+            } else {
+                return;
+            }
+            checkAnnotationsInClass(elt, field, varTree, getCurrentPath(), kind);
+        }
+    }
+
+    private void checkAnnotationsInClass(
+            Element elt,
+            Receiver canonicalTargetName,
+            Tree tree,
+            TreePath path,
+            SideEffectKind sideEffectKind) {
+        List<? extends Element> enclosedElts =
+                ElementUtils.enclosingClass(elt).getEnclosedElements();
+        List<AnnotatedTypeMirror> enclosedTypes = findEnclosedTypes(enclosedElts);
+        for (AnnotatedTypeMirror atm : enclosedTypes) {
+            List<Receiver> rs =
+                    UpperBoundUtil.getDependentReceivers(atm.getAnnotations(), path, atypeFactory);
+            for (Receiver r : rs) {
+                SideEffectError result =
+                        UpperBoundUtil.isSideEffected(r, canonicalTargetName, sideEffectKind);
+                if (result != UpperBoundUtil.SideEffectError.NO_ERROR) {
+                    checker.report(Result.failure(result.errorKey, atm), tree);
+                }
+            }
+        }
+    }
+
+    private List<AnnotatedTypeMirror> findEnclosedTypes(List<? extends Element> enclosedElts) {
+        List<AnnotatedTypeMirror> enclosedTypes = new ArrayList<>();
+        for (Element e : enclosedElts) {
+            AnnotatedTypeMirror atm = atypeFactory.getAnnotatedType(e);
+            enclosedTypes.add(atm);
+            if (e.getKind() == ElementKind.METHOD) {
```

Comment:
What about enclosed class or interfaces?

Reply:
Would those have more than a single annotated type associated with them? Methods have parameter and return types that need to be calculated separately, but classes/interfaces should only have the annotation that's written on the declaration.

Actually, I might need to handle generic bounds here. Is that what you mean?

Reply:
I mean methods in an enclosed class or interface.

Reply:
I was under the impression that `enclosedElts` should already contain those methods.

Reply:
Oh and anonymous classes?

Reply:
>I was under the impression that enclosedElts should already contain those methods.

From `getEnclosedElements` Javadoc:

>Returns the fields, methods, constructors, and member types that are directly declared in this class or interface.

Reply:
So do you think I should be using `Elements#getAllMembers` instead? That takes a `TypeElement`; could I pass the result of `ElementUtils#getEnclosingClass` to that method? If not (I suspect not), how would I go about making the conversion?

Reply:
A method of an inner class isn't a member of the outer class, so `Elements#getAllMembers` won't work either.  I would recur in the the case of an inner class.

Reply:
Also, `ElementUtils#getEnclosingClass` returns a `TypeElement`.

Reply:
This should work:
```
            } else if(e.getKind().isClass()) {
                enclosedTypes.addAll(findEnclosedTypes(e.getEnclosedElements()));
            }
```
Plus add a test case.

Reply:
Alright, will do. Thanks.